File: sdk/go/query.go
// The typed query builder.
//
// Chainable in Go, because that is what makes a query pleasant to write. What it
// produces is an AST, and the AST is rendered to SQL-subset source and bound
// parameters by the WebAssembly core — so the same query built in TypeScript or
// Python reaches the server as the same bytes, and the rule that a value never
// becomes query text has one implementation rather than one per language.
//
// Nothing here interpolates. `Where(Eq("name", userInput))` puts `userInput` in
// the parameter map and `$p0` in the text, whatever `userInput` contains.

package thetabase

// CompareOp is a binary comparison. The set is closed on purpose: an open one
// would eventually carry an operator the renderer does not know, and the only
// place left to put it would be the query text.
type CompareOp string

const (
	OpEq  CompareOp = "eq"
	OpNe  CompareOp = "ne"
	OpLt  CompareOp = "lt"
	OpLte CompareOp = "lte"
	OpGt  CompareOp = "gt"
	OpGte CompareOp = "gte"
)

// Predicate is one node of a filter tree.
//
// One struct with a Kind rather than an interface with an implementation per
// arm, because the AST crosses a JSON boundary into the core. An interface would
// need custom marshalling on both sides, and the shape that survives the round
// trip unchanged is the one the other two SDKs already send.
type Predicate struct {
	Kind   string      `json:"kind"`
	Column string      `json:"column,omitempty"`
	Op     CompareOp   `json:"op,omitempty"`
	Value  any         `json:"value,omitempty"`
	Values []any       `json:"values,omitempty"`
	Terms  []Predicate `json:"terms,omitempty"`
	Term   *Predicate  `json:"term,omitempty"`
}

// OrderBy is one sort key.
type OrderBy struct {
	Column     string `json:"column"`
	Descending bool   `json:"descending"`
}

// QueryAst is what the core renders. Exported so a caller can inspect or send
// one without going through the builder.
type QueryAst struct {
	Table   string     `json:"table"`
	Columns []string   `json:"columns"`
	Filter  *Predicate `json:"filter,omitempty"`
	OrderBy []OrderBy  `json:"orderBy"`
	Limit   *int       `json:"limit,omitempty"`
	Offset  *int       `json:"offset,omitempty"`
}

func compare(column string, op CompareOp, value any) Predicate {
	return Predicate{Kind: "compare", Column: column, Op: op, Value: value}
}

// Eq keeps rows where `column` equals `value`.
func Eq(column string, value any) Predicate { return compare(column, OpEq, value) }

// Ne keeps rows where `column` does not equal `value`.
func Ne(column string, value any) Predicate { return compare(column, OpNe, value) }

// Lt keeps rows where `column` is less than `value`.
func Lt(column string, value any) Predicate { return compare(column, OpLt, value) }

// Lte keeps rows where `column` is at most `value`.
func Lte(column string, value any) Predicate { return compare(column, OpLte, value) }

// Gt keeps rows where `column` is greater than `value`.
func Gt(column string, value any) Predicate { return compare(column, OpGt, value) }

// Gte keeps rows where `column` is at least `value`.
func Gte(column string, value any) Predicate { return compare(column, OpGte, value) }

// In keeps rows where `column` is one of `values`.
func In(column string, values []any) Predicate {
	return Predicate{Kind: "in", Column: column, Values: values}
}

// IsNull keeps rows where `column` is null.
func IsNull(column string) Predicate { return Predicate{Kind: "isNull", Column: column} }

// NotNull keeps rows where `column` is not null.
func NotNull(column string) Predicate { return Predicate{Kind: "notNull", Column: column} }

// And keeps rows matching every term.
func And(terms ...Predicate) Predicate { return Predicate{Kind: "and", Terms: terms} }

// Or keeps rows matching any term.
func Or(terms ...Predicate) Predicate { return Predicate{Kind: "or", Terms: terms} }

// Not inverts a term.
func Not(term Predicate) Predicate { return Predicate{Kind: "not", Term: &term} }

// Query is a typed plan under construction.
type Query struct {
	ast QueryAst
}

// Table starts a query against `name`.
func Table(name string) Query {
	return Query{ast: QueryAst{Table: name, Columns: []string{}, OrderBy: []OrderBy{}}}
}

// Select adds columns to the projection.
//
// Every method here returns a new Query rather than mutating the receiver. A
// builder that mutated would make a shared base query change under whoever else
// was holding it — and Go's value semantics do not save it, because the slices
// inside would still be shared.
func (q Query) Select(columns ...string) Query {
	next := q.clone()
	next.ast.Columns = append(next.ast.Columns, columns...)
	return next
}

// Where keeps rows matching `predicate`. Repeated calls are ANDed.
func (q Query) Where(predicate Predicate) Query {
	next := q.clone()
	if next.ast.Filter != nil {
		combined := And(*next.ast.Filter, predicate)
		next.ast.Filter = &combined
	} else {
		next.ast.Filter = &predicate
	}
	return next
}

// OrderBy adds a sort key.
func (q Query) OrderBy(column string, descending bool) Query {
	next := q.clone()
	next.ast.OrderBy = append(next.ast.OrderBy, OrderBy{Column: column, Descending: descending})
	return next
}

// Limit caps the number of rows returned.
func (q Query) Limit(count int) Query {
	next := q.clone()
	next.ast.Limit = &count
	return next
}

// Offset skips rows before returning any.
func (q Query) Offset(count int) Query {
	next := q.clone()
	next.ast.Offset = &count
	return next
}

// Ast returns the plan, for a caller that wants to inspect or send it.
func (q Query) Ast() QueryAst { return q.clone().ast }

// Render renders through the core. Fails if an identifier could carry syntax.
func (q Query) Render(core *ScribeCore) (RenderedQuery, error) {
	return core.RenderQuery(q.ast)
}

// clone deep-copies the parts a later call would otherwise append into.
//
// `append` on a slice with spare capacity writes in place, so two queries
// derived from one base would share — and the second `Select` would overwrite
// the first's column rather than adding to its own.
func (q Query) clone() Query {
	next := q
	// `make` + `copy` rather than `append(nil, ...)`: appending nothing to a nil
	// slice yields nil, and a nil slice marshals to `null` where an empty one
	// marshals to `[]`. The core reads `columns` as a sequence and refuses null,
	// so every query with no explicit projection — the common case — failed to
	// render. Found by the conformance harness on the first Go run, which is
	// exactly the divergence it exists to catch.
	next.ast.Columns = make([]string, len(q.ast.Columns))
	copy(next.ast.Columns, q.ast.Columns)
	next.ast.OrderBy = make([]OrderBy, len(q.ast.OrderBy))
	copy(next.ast.OrderBy, q.ast.OrderBy)
	if q.ast.Filter != nil {
		filter := *q.ast.Filter
		next.ast.Filter = &filter
	}
	if q.ast.Limit != nil {
		limit := *q.ast.Limit
		next.ast.Limit = &limit
	}
	if q.ast.Offset != nil {
		offset := *q.ast.Offset
		next.ast.Offset = &offset
	}
	return next
}

File: sdk/rust/src/query.rs
//! The typed query builder.
//!
//! Chainable, because that is what makes a query pleasant to write. What it
//! produces is a [`QueryAst`], and the AST is rendered to SQL-subset source and
//! bound parameters by the Scribe core — the same renderer the TypeScript,
//! Python and Go builders reach through WebAssembly. A query built the same way
//! in any of them arrives as the same bytes, and the rule that a value never
//! becomes query text has one implementation.
//!
//! Nothing here interpolates. `filter(eq("name", user_input))` puts
//! `user_input` in the parameter map and `$p0` in the text, whatever
//! `user_input` contains.
//!
//! # `filter`, not `where`
//!
//! `where` is a Rust keyword, so the method that the other three SDKs spell
//! `where` is `filter` here. Named rather than escaped as `r#where`: a raw
//! identifier at every call site is a worse tax than one method name that reads
//! differently, and `filter` is what a Rust caller would reach for anyway.

use serde_json::Value;
use theta_scribe_wasm::query::{CompareOp, Predicate, QueryAst, Sort};

pub use theta_scribe_wasm::query::{RenderError, Rendered};

fn compare(column: &str, op: CompareOp, value: impl Into<Value>) -> Predicate {
    Predicate::Compare {
        column: column.to_string(),
        op,
        value: value.into(),
    }
}

/// Keep rows where `column` equals `value`.
pub fn eq(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Eq, value)
}

/// Keep rows where `column` does not equal `value`.
pub fn ne(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Ne, value)
}

/// Keep rows where `column` is less than `value`.
pub fn lt(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Lt, value)
}

/// Keep rows where `column` is at most `value`.
pub fn lte(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Lte, value)
}

/// Keep rows where `column` is greater than `value`.
pub fn gt(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Gt, value)
}

/// Keep rows where `column` is at least `value`.
pub fn gte(column: &str, value: impl Into<Value>) -> Predicate {
    compare(column, CompareOp::Gte, value)
}

/// Keep rows where `column` is one of `values`.
pub fn is_in(column: &str, values: impl IntoIterator<Item = Value>) -> Predicate {
    Predicate::In {
        column: column.to_string(),
        values: values.into_iter().collect(),
    }
}

/// Keep rows where `column` is null.
pub fn is_null(column: &str) -> Predicate {
    Predicate::IsNull {
        column: column.to_string(),
    }
}

/// Keep rows where `column` is not null.
pub fn not_null(column: &str) -> Predicate {
    Predicate::NotNull {
        column: column.to_string(),
    }
}

/// Keep rows matching every term.
pub fn and(terms: impl IntoIterator<Item = Predicate>) -> Predicate {
    Predicate::And {
        terms: terms.into_iter().collect(),
    }
}

/// Keep rows matching any term.
pub fn or(terms: impl IntoIterator<Item = Predicate>) -> Predicate {
    Predicate::Or {
        terms: terms.into_iter().collect(),
    }
}

/// Invert a term.
pub fn not(term: Predicate) -> Predicate {
    Predicate::Not {
        term: Box::new(term),
    }
}

/// Start a query against `name`.
pub fn table(name: &str) -> Query {
    Query {
        ast: QueryAst {
            table: name.to_string(),
            columns: Vec::new(),
            filter: None,
            order_by: Vec::new(),
            limit: None,
            offset: None,
        },
    }
}

/// A typed plan under construction.
///
/// Owned and consumed by each method rather than cloned, which is what Rust
/// offers and the other SDKs cannot: a builder that returned a new value from a
/// borrow would copy the AST on every call, and one that mutated in place would
/// let a shared base query change under whoever else held it. Moving makes the
/// second impossible at compile time.
#[derive(Debug)]
pub struct Query {
    ast: QueryAst,
}

impl Query {
    /// Add columns to the projection. No projection means every column.
    pub fn select(mut self, columns: impl IntoIterator<Item = impl Into<String>>) -> Self {
        self.ast.columns.extend(columns.into_iter().map(Into::into));
        self
    }

    /// Keep rows matching `predicate`. Repeated calls are ANDed.
    pub fn filter(mut self, predicate: Predicate) -> Self {
        self.ast.filter = Some(match self.ast.filter.take() {
            Some(existing) => Predicate::And {
                terms: vec![existing, predicate],
            },
            None => predicate,
        });
        self
    }

    /// Add a sort key.
    pub fn order_by(mut self, column: &str, descending: bool) -> Self {
        self.ast.order_by.push(Sort {
            column: column.to_string(),
            descending,
        });
        self
    }

    /// Cap the number of rows returned.
    pub fn limit(mut self, count: u64) -> Self {
        self.ast.limit = Some(count);
        self
    }

    /// Skip rows before returning any.
    pub fn offset(mut self, count: u64) -> Self {
        self.ast.offset = Some(count);
        self
    }

    /// The plan, for a caller that wants to inspect it.
    pub fn ast(&self) -> &QueryAst {
        &self.ast
    }

    /// Render to SQL-subset source and bound parameters.
    ///
    /// Fails if an identifier could carry syntax. Table and column names are
    /// written into the query text, so they are restricted to letters, digits
    /// and underscores — a name that needed quoting would be a way to write
    /// syntax, and the renderer refuses rather than escaping.
    pub fn render(&self) -> Result<Rendered, RenderError> {
        theta_scribe_wasm::query::render(&self.ast)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_value_never_reaches_the_query_text() {
        // The property the whole builder exists for.
        let hostile = "'; DROP TABLE users; --";
        let rendered = table("users")
            .filter(eq("name", hostile))
            .render()
            .expect("renders");

        assert!(
            !rendered.sql.contains("DROP"),
            "the value reached the text: {}",
            rendered.sql
        );
        assert_eq!(rendered.sql, "SELECT * FROM users WHERE name = $p0");
        assert_eq!(
            rendered.params.get("p0").map(String::as_str),
            Some("\"'; DROP TABLE users; --\"")
        );
    }

    #[test]
    fn repeated_filters_are_anded_rather_than_replaced() {
        let rendered = table("users")
            .filter(eq("email", "a@example.com"))
            .filter(gt("age", 30))
            .render()
            .expect("renders");
        assert!(
            rendered.sql.contains(" AND "),
            "the second filter replaced the first: {}",
            rendered.sql
        );
        assert_eq!(rendered.params.len(), 2);
    }

    #[test]
    fn an_identifier_that_could_carry_syntax_is_refused() {
        // Refused rather than escaped. Escaping would mean deciding what is safe
        // to quote, and the safe answer is that a name needing quotes is not a
        // name this subset accepts.
        let refused = table("users; DROP TABLE x").render();
        assert!(matches!(refused, Err(RenderError::BadIdentifier(_))));
    }

    #[test]
    fn a_query_with_no_projection_selects_everything() {
        // Go's binding got this wrong in a way Rust cannot: an empty `Vec` is an
        // empty `Vec`, where an empty Go slice can be nil and marshal to `null`.
        // Kept as a test anyway so the two SDKs are known to agree here.
        let rendered = table("users").render().expect("renders");
        assert_eq!(rendered.sql, "SELECT * FROM users");
    }

    #[test]
    fn the_builder_moves_rather_than_sharing() {
        // Compile-time proof that a base query cannot change under a holder: the
        // AST is owned, so two derived queries cannot exist over one buffer the
        // way Go's slices can. What this asserts is the visible half — that the
        // derived query carries both columns.
        let rendered = table("users")
            .select(["id"])
            .select(["email"])
            .render()
            .expect("renders");
        assert_eq!(rendered.sql, "SELECT id, email FROM users");
    }
}
